File: internal/plugins/jest/rules/no_focused_tests/no_focused_tests_test.go
package no_focused_tests_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_focused_tests"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoFocusedTests(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_focused_tests.NoFocusedTestsRule,
		[]rule_tester.ValidTestCase{
			{Code: "import { describe as fdescribe } from '@jest/globals';\n\nfdescribe()"},
			{Code: "describe()"},
			{Code: "it()"},
			{Code: "describe.skip()"},
			{Code: "it.skip()"},
			{Code: "test()"},
			{Code: "test.skip()"},
			{Code: "var appliedOnly = describe.only; appliedOnly.apply(describe)"},
			{Code: "var calledOnly = it.only; calledOnly.call(it)"},
			{Code: "it.each()()"},
			{Code: "it.each`table`()"},
			{Code: "test.each()()"},
			{Code: "test.each`table`()"},
			{Code: "test.concurrent()"},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: "describe.only()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    10,
						EndLine:   1,
						EndColumn: 14,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "describe()",
							},
						},
					},
				},
			},

			// TODO: Fix this test case
			// Currently, rslint does not support pragma configuration
			// {
			// 	Code: "context.only()",
			// 	Errors: []rule_tester.InvalidTestCaseError{
			// 		{
			// 			MessageId: "focusedTest",
			// 			Line:      1,
			// 			Column:    9,
			// 			EndLine:   1,
			// 			EndColumn: 13,
			// 			Suggestions: []rule_tester.InvalidTestCaseSuggestion{
			// 				{
			// 					MessageId: "suggestRemoveFocus",
			// 					Output:    "context()",
			// 				},
			// 			},
			// 		},
			// 	},
			// },

			{
				Code: "describe.only.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    10,
						EndLine:   1,
						EndColumn: 14,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "describe.each()()",
							},
						},
					},
				},
			},
			{
				Code: "describe.only.each`table`()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    10,
						EndLine:   1,
						EndColumn: 14,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "describe.each`table`()",
							},
						},
					},
				},
			},
			{
				Code: `describe["only"]()`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    10,
						EndLine:   1,
						EndColumn: 16,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "describe()",
							},
						},
					},
				},
			},
			{
				Code: "it.only()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    4,
						EndLine:   1,
						EndColumn: 8,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it()",
							},
						},
					},
				},
			},
			{
				Code: "it.concurrent.only.each``()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    15,
						EndLine:   1,
						EndColumn: 19,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it.concurrent.each``()",
							},
						},
					},
				},
			},
			{
				Code: "it.only.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    4,
						EndLine:   1,
						EndColumn: 8,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it.each()()",
							},
						},
					},
				},
			},
			{
				Code: "it.only.each`table`()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    4,
						EndLine:   1,
						EndColumn: 8,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it.each`table`()",
							},
						},
					},
				},
			},
			{
				Code: `it["only"]()`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    4,
						EndLine:   1,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it()",
							},
						},
					},
				},
			},
			{
				Code: "test.only()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    6,
						EndLine:   1,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "test()",
							},
						},
					},
				},
			},
			{
				Code: "test.concurrent.only.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    17,
						EndLine:   1,
						EndColumn: 21,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "test.concurrent.each()()",
							},
						},
					},
				},
			},
			{
				Code: "test.only.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    6,
						EndLine:   1,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "test.each()()",
							},
						},
					},
				},
			},
			{
				Code: "test.only.each`table`()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    6,
						EndLine:   1,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "test.each`table`()",
							},
						},
					},
				},
			},
			{
				Code: `test["only"]()`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    6,
						EndLine:   1,
						EndColumn: 12,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "test()",
							},
						},
					},
				},
			},
			{
				Code: "fdescribe()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    1,
						EndLine:   1,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "describe()",
							},
						},
					},
				},
			},
			{
				Code: "fit()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      1,
						Column:    1,
						EndLine:   1,
						EndColumn: 4,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it()",
							},
						},
					},
				},
			},
			{
			Code: "fit.each()()",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "focusedTest",
					Line:      1,
					Column:    1,
					EndLine:   1,
					EndColumn: 4,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it.each()()",
							},
						},
					},
				},
			},
			{
			Code: "fit.each`table`()",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "focusedTest",
					Line:      1,
					Column:    1,
					EndLine:   1,
					EndColumn: 4,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "it.each`table`()",
							},
						},
					},
				},
			},
			{
				Code: "const { describe } = require('@jest/globals');\n\ndescribe.only()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      3,
						Column:    10,
						EndLine:   3,
						EndColumn: 14,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "const { describe } = require('@jest/globals');\n\ndescribe()",
							},
						},
					},
				},
			},
			{
				Code: "import { describe as describeThis } from '@jest/globals';\n\ndescribeThis.only()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      3,
						Column:    14,
						EndLine:   3,
						EndColumn: 18,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "import { describe as describeThis } from '@jest/globals';\n\ndescribeThis()",
							},
						},
					},
				},
			},
			{
				Code: "const { fdescribe } = require('@jest/globals');\n\nfdescribe()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      3,
						Column:    1,
						EndLine:   3,
						EndColumn: 10,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "const { fdescribe } = require('@jest/globals');\n\ndescribe()",
							},
						},
					},
				},
			},
			{
				Code: "import { fdescribe as describeJustThis } from '@jest/globals';\n\ndescribeJustThis()\ndescribeJustThis.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      3,
						Column:    1,
						EndLine:   3,
						EndColumn: 17,
					},
					{
						MessageId: "focusedTest",
						Line:      4,
						Column:    1,
						EndLine:   4,
						EndColumn: 24,
					},
				},
			},
			{
				Code: "import { describe as context } from '@jest/globals';\n\ncontext.only.each()()",
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "focusedTest",
						Line:      3,
						Column:    9,
						EndLine:   3,
						EndColumn: 13,
						Suggestions: []rule_tester.InvalidTestCaseSuggestion{
							{
								MessageId: "suggestRemoveFocus",
								Output:    "import { describe as context } from '@jest/globals';\n\ncontext.each()()",
							},
						},
					},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-focused-tests.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-focused-tests', {} as never, {
  valid: [
    { code: 'describe()' },
    { code: 'it()' },
    { code: 'describe.skip()' },
    { code: 'it.skip()' },
    { code: 'test()' },
    { code: 'test.skip()' },
    { code: 'var appliedOnly = describe.only; appliedOnly.apply(describe)' },
    { code: 'var calledOnly = it.only; calledOnly.call(it)' },
    { code: 'it.each()()' },
    { code: 'it.each`table`()' },
    { code: 'test.each()()' },
    { code: 'test.each`table`()' },
    { code: 'test.concurrent()' },
  ],
  invalid: [
    {
      code: 'describe.only()',
      errors: [
        {
          line: 1,
          column: 10,
          endLine: 1,
          endColumn: 14,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'describe()',
            },
          ],
        },
      ],
    },

    // TODO: Fix this test case
    // {
    //   code: 'context.only()',
    //   errors: [
    //     {
    //       line: 1,
    //       column: 9,
    //       endLine: 1,
    //       endColumn: 13,
    //       messageId: 'focusedTest',
    //       suggestions: [
    //         {
    //           messageId: 'suggestRemoveFocus',
    //           output: 'context()',
    //         },
    //       ],
    //     },
    //   ],
    //   settings: { jest: { globalAliases: { describe: ['context'] } } },
    // },

    {
      code: 'describe.only.each()()',
      errors: [
        {
          line: 1,
          column: 10,
          endLine: 1,
          endColumn: 14,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'describe.each()()',
            },
          ],
        },
      ],
    },
    {
      code: 'describe.only.each`table`()',
      errors: [
        {
          line: 1,
          column: 10,
          endLine: 1,
          endColumn: 14,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'describe.each`table`()',
            },
          ],
        },
      ],
    },
    {
      code: 'describe["only"]()',
      errors: [
        {
          line: 1,
          column: 10,
          endLine: 1,
          endColumn: 16,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'describe()',
            },
          ],
        },
      ],
    },
    {
      code: 'it.only()',
      errors: [
        {
          line: 1,
          column: 4,
          endLine: 1,
          endColumn: 8,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it()',
            },
          ],
        },
      ],
    },
    {
      code: 'it.concurrent.only.each``()',
      errors: [
        {
          line: 1,
          column: 15,
          endLine: 1,
          endColumn: 19,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it.concurrent.each``()',
            },
          ],
        },
      ],
    },
    {
      code: 'it.only.each()()',
      errors: [
        {
          line: 1,
          column: 4,
          endLine: 1,
          endColumn: 8,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it.each()()',
            },
          ],
        },
      ],
    },
    {
      code: 'it.only.each`table`()',
      errors: [
        {
          line: 1,
          column: 4,
          endLine: 1,
          endColumn: 8,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it.each`table`()',
            },
          ],
        },
      ],
    },
    {
      code: 'it["only"]()',
      errors: [
        {
          line: 1,
          column: 4,
          endLine: 1,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it()',
            },
          ],
        },
      ],
    },
    {
      code: 'test.only()',
      errors: [
        {
          line: 1,
          column: 6,
          endLine: 1,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'test()',
            },
          ],
        },
      ],
    },
    {
      code: 'test.concurrent.only.each()()',
      errors: [
        {
          line: 1,
          column: 17,
          endLine: 1,
          endColumn: 21,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'test.concurrent.each()()',
            },
          ],
        },
      ],
    },
    {
      code: 'test.only.each()()',
      errors: [
        {
          line: 1,
          column: 6,
          endLine: 1,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'test.each()()',
            },
          ],
        },
      ],
    },
    {
      code: 'test.only.each`table`()',
      errors: [
        {
          line: 1,
          column: 6,
          endLine: 1,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'test.each`table`()',
            },
          ],
        },
      ],
    },
    {
      code: 'test["only"]()',
      errors: [
        {
          line: 1,
          column: 6,
          endLine: 1,
          endColumn: 12,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'test()',
            },
          ],
        },
      ],
    },
    {
      code: 'fdescribe()',
      errors: [
        {
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'describe()',
            },
          ],
        },
      ],
    },
    {
      code: 'fit()',
      errors: [
        {
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 4,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it()',
            },
          ],
        },
      ],
    },
    {
      code: 'fit.each()()',
      errors: [
        {
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 4,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it.each()()',
            },
          ],
        },
      ],
    },
    {
      code: 'fit.each`table`()',
      errors: [
        {
          line: 1,
          column: 1,
          endLine: 1,
          endColumn: 4,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: 'it.each`table`()',
            },
          ],
        },
      ],
    },
  ],
});

// import test cases
ruleTester.run('no-focused-tests', {} as never, {
  valid: [
    {
      code: `
        import { describe as fdescribe } from '@jest/globals';

        fdescribe()
      `,
    },
  ],

  invalid: [
    {
      code: `
        const { describe } = require('@jest/globals');

        describe.only()
      `,
      errors: [
        {
          line: 3,
          column: 10,
          endLine: 3,
          endColumn: 14,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: `
                const { describe } = require('@jest/globals');

                describe()
              `,
            },
          ],
        },
      ],
    },
    {
      code: `
        import { describe as describeThis } from '@jest/globals';

        describeThis.only()
      `,
      errors: [
        {
          line: 3,
          column: 14,
          endLine: 3,
          endColumn: 18,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: `
                import { describe as describeThis } from '@jest/globals';

                describeThis()
              `,
            },
          ],
        },
      ],
    },
    {
      code: `
        const { fdescribe } = require('@jest/globals');

        fdescribe()
      `,
      errors: [
        {
          line: 3,
          column: 1,
          endLine: 3,
          endColumn: 10,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: `
                const { fdescribe } = require('@jest/globals');

                describe()
              `,
            },
          ],
        },
      ],
    },
  ],
});

// alias test cases
ruleTester.run('no-focused-tests', {} as never, {
  valid: [],

  invalid: [
    {
      code: `
        import { describe as describeThis } from '@jest/globals';

        describeThis.only()
      `,
      errors: [
        {
          line: 3,
          column: 14,
          endLine: 3,
          endColumn: 18,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: `
                import { describe as describeThis } from '@jest/globals';

                describeThis()
              `,
            },
          ],
        },
      ],
    },
    {
      code: `
        import { fdescribe as describeJustThis } from '@jest/globals';

        describeJustThis()
        describeJustThis.each()()
      `,
      errors: [
        {
          line: 3,
          column: 1,
          endLine: 3,
          endColumn: 17,
          messageId: 'focusedTest',
        },
        {
          line: 4,
          column: 1,
          endLine: 4,
          endColumn: 17,
          messageId: 'focusedTest',
        },
      ],
    },
    {
      code: `
        import { describe as context } from '@jest/globals';

        context.only.each()()
      `,
      errors: [
        {
          line: 3,
          column: 9,
          endLine: 3,
          endColumn: 13,
          messageId: 'focusedTest',
          suggestions: [
            {
              messageId: 'suggestRemoveFocus',
              output: `
                import { describe as context } from '@jest/globals';

                context.each()()
              `,
            },
          ],
        },
      ],
    },
  ],
});
